### ISSUE TITLE
ERC20_base initializer

### ISSUE THREAD

**Issue by calvwang9** [OpenZeppelin/cairo-contracts#194]
**🧐 Motivation**

While experimenting with the ERC20_base contract, I noticed some interesting design choices that were slightly unintuitive to me. Hoping to open a conversation and better understand the rationale behind these features and suggest some alternatives 😄 

**📝 Details**

The ERC20_base initializer currently looks like this:
```
func ERC20_initializer{
        syscall_ptr : felt*, 
        pedersen_ptr : HashBuiltin*,
        range_check_ptr
    }(
        name: felt,
        symbol: felt,
        initial_supply: Uint256,
        recipient: felt
    ):
    ERC20_name_.write(name)
    ERC20_symbol_.write(symbol)
    ERC20_decimals_.write(18)
    ERC20_mint(recipient, initial_supply)
    return ()
end
```
Notably:
1. Decimals is hardcoded to 18
   - While 18 decimals has become a standard for most ERC20s, we should still be able to define this variable to be otherwise. For example, stablecoins may want fewer decimals, like USDC (6), and other ERC20 use cases (e.g. in-game resources) may require less or more granularity
   - The Solidity ERC20 implementation sets decimals to 18 by default but also allows overriding in child contracts, unlike Cairo. In favor of creating a generalizable base contract for derived contracts to use out of the box, could decimals be defined as a parameter in the constructor instead?

2. An initial supply of tokens is minted to a recipient upon initialization
   - This could be convenient for few but is likely unnecessary for many ERC20 implementations. The benefits of saving an additional transaction to mint an initial supply is fairly insignificant especially in Cairo with little or no transaction fees.
   - The Solidity ERC20 implementation takes an agnostic approach to token supply mechanisms and just provides the internal `_mint` function for derived contracts to use in any way they choose, any reasons against using the same pattern?

An example change to the initializer in `ERC20_base.cairo` to address the above points:
```
func ERC20_initializer{
        syscall_ptr : felt*, 
        pedersen_ptr : HashBuiltin*,
        range_check_ptr
    }(
        name: felt,
        symbol: felt,
        decimals: felt,
    ):
    ERC20_name_.write(name)
    ERC20_symbol_.write(symbol)
    ERC20_decimals_.write(decimals)
    return ()
end
```

I believe this allows users to use the base ERC20 contract with minimal modifications for most ERC20 implementations, though happy to discuss and explore any counter-arguments or things I may have missed!

**Comment by martriay** [comment]
Hello @calvwang9! I don't think there's much to discuss in here. I fully agree with all of your points. My only addition would be to check that decimals are under 2^8 as specified in ERC20.

Do you want to open a PR with the changes?

**Comment by andrew-fleming** [comment]
Resolved in #195
